### DISCUSSION TITLE
Apply new parameterization (UV) to already textured 3D mesh in order to flatten texture to 2D

### DISCUSSION THREAD

**Discussion by HansVansWardikins** [pyvista/pyvista#5007] [Q&A]
Hello, 

I am looking to use Pyvista to apply an As-Rigid-As-Possible (ARAP) flattening to a textured mesh. The application is for calibration of a simulation for a plastic deformation process (thats a mouthful!). I am using LibIGL to flatten my mesh using the ARAP algorithm, however, LibIGL doesnt seem to support textured mesh's, thus looking to Pyvista. The output from the LibIGL ARAP function is a UV which maps every mesh vertex to a 2D plane in a way that minimizes face deformation. I want to then be able to apply that UV map to an already textured mesh in order to flatten it.

This is the textured model as output from mesh.plot(texture=tex) after being imported from a .obj file with vertex texture coordinates (vt) and a .png texture image.
![GetImage](https://github.com/pyvista/pyvista/assets/145659314/eceabe86-cd25-4cc0-9fc5-41c6baaccc94)

These are the outputs from the LibIGL ARAP function, the 3D mesh displayed with an arbitrary checkerboard and the 2D flattened mesh once the faces have been mapped to a 2D plane using the ARAP output UV:
![GetImage (1)](https://github.com/pyvista/pyvista/assets/145659314/3a59e041-5ef5-40fb-8dcc-57a3cb3df344)
![GetImage (2)](https://github.com/pyvista/pyvista/assets/145659314/1808fedf-a9cf-420c-897a-5813744e6204)

Now I want to be able to take the mesh as textured in the first image and apply the ARAP output UV to in in order to achieve a textured 2d representation of the mesh as such:
![GetImage (3)](https://github.com/pyvista/pyvista/assets/145659314/cf8f5871-8bab-4ee3-b763-d7253c1fdf50)

It seems like I need to be able to store each vertex with its color information as displayed by mesh.plot(texture=tex) and then replot it with the output 2D UV array rather than the 3D face array from the .obj file.

I would really appreciate it if anyone has ideas on how I may achieve this. So far I am struggling to find any function that is able to achieve this.

Best regards
Alex

**Comment by darikg** [comment]
Apologies if I'm misunderstanding but shouldn't it be as simple as

```python
uv = igl.arap(...)
mesh.points = np.stack([uv, zeros((mesh.n_points, 1))], axis=1)
mesh.plot(texture=tex)
```
?

**Reply by HansVansWardikins** [reply]
Ah yes! Thinking about it you are exactly right, amazing how I could over complicate the problem! 

One thing I am find though is that the result doesn't seem to render correctly...  I did use `mesh.points = np.hstack((uv, np.zeros((mesh.n_points, 1))))` instead of your suggestion `np.stack` as stack seemed to only work with arrays of equal shape.

After finding it didn't render correctly I tried manually editing the .obj file, replacing the `v` coords with the uv coords and zeros in the z column, I did this by exporting a CSV for the uv and editing it into the .obj with Excel. When I import the edited .obj with PyVista it renders correctly! How strange!

Any ideas why this might be?

Using:
mesh.points = np.hstack((uva, np.zeros((mesh.n_points, 1))))
tex = pv.read_texture("model.jpg")
mesh.plot(texture=tex):
![UVA broken render](https://github.com/pyvista/pyvista/assets/145659314/37862d14-3782-42c8-be30-044d3524f00b)


Using:
mesh = pv.read("excel-recompiled-model-uva.obj")
tex = pv.read_texture("model.jpg")
mesh.plot(texture=tex):
![Flattened with excel](https://github.com/pyvista/pyvista/assets/145659314/fd3e1aaf-b058-4db9-9b28-f74d6303bf75)

**Reply by darikg** [reply]
> One thing I am find though is that the result doesn't seem to render correctly... I did use mesh.points = np.hstack((uv, np.zeros((mesh.n_points, 1)))) instead of your suggestion np.stack as stack seemed to only work with arrays of equal shape.

Sorry! I always get that wrong the first time. I think `np.concatenate([uv, zeros((mesh.n_points, 1))], axis=1)` would've worked too.

> Any ideas why this might be?

That's really weird! My first thought was some sort of row major / column major order incompatibility between pyvista and igl causing the point coordinates to get scrambled but from that gif all the points look correctly positioned at z=0 and the border looks correct so I guess the faces array instead is getting scrambled somehow. How are you passing the faces array to/from igl?

**Reply by HansVansWardikins** [reply]
I am passing the faces to IGL with `v, f = igl.read_triangle_mesh("Model.obj")`.

I have found a couple of suspicious things but am yet to figure out exactly whats going on:

### First suspicious thing, the number of vertices:
- When the model is read in with the above line, IGL finds 38353 vertices and 75917 faces and the arrays correctly match the .obj file.
- When the model is read in with `mesh = pv.read("Model.obj")`, `mesh.points.shape[0]` returns 227751, even though there is definitely only 38353 v lines in the .obj file?? `mesh.number_of_cells` returns 75917 as expected.
- When `mesh = mesh.clean()` is called after the .obj is read, `mesh.points.shape[0]` then returns 38353, and `mesh.number_of_cells` still returns 75917 as expected, both inline with the IGL import...
- So in order to perform the `mesh.points = np.hstack((uva, np.zeros((mesh.n_points, 1))))` operation previously I had been using `mesh = mesh.clean()` to ensure the number of rows in uva and mesh.points matched so that the hstack operation could be performed.
- Interestingly when the 3D model is plotted after `mesh = mesh.clean()` some strange artifacts can be seen:
![image](https://github.com/pyvista/pyvista/assets/145659314/cbd5e07b-f5ca-413b-8f77-bef0d8866a0c)

- Also when using the file I manually manipulated with Excel, `mesh = pv.read("excel-recompiled-model-uva-flattened.obj")` `mesh.points.shape[0]` returns 227751, and after using `mesh = mesh.clean()` the same artifacts are seen on the 2D texture:
![image](https://github.com/pyvista/pyvista/assets/145659314/67ca7d02-851c-4ff4-880f-4ebc5c8c3997)

### Second suspicious thing, all 3 arrays don't match original file after clean:

- `mesh = mesh.clean()` seems to completely rebuild all of the arrays with new coordinates...

### AHA HERE'S MY PROBLEM!!

Because I have to use `mesh = mesh.clean()` to get the lengths of **mesh.points** and **uva** to match and because **uva** is no longer a proper reference for the new face array created by  `mesh = mesh.clean()` the texture is incorrectly applied.

So the solution should be to match the length of **mesh.points** and **uva** by a means other than `mesh.clean()` and pass IGL and PyVista the same arrays. I borrowed a manual .obj parser from [HERE](https://github.com/pyvista/pyvista-support/issues/230?fbclid=IwAR3L87yVSZXrVnHPxeVnaVobBhl00b1jfU-kxRxfOxrxli_1XF2rKL6TTCY) thanks @banesullivan, however I am now having an issue with the lengths of **v** and **vt** not matching, I am getting **vt** as 39739 rows and **v** as 38353 rows, this means that when passing PyVista the arrays with: 

![image](https://github.com/pyvista/pyvista/assets/145659314/aa3eebe0-ad6d-4b12-8d74-4bb8f388531b)

I get:
![image](https://github.com/pyvista/pyvista/assets/145659314/adc32def-2f70-4ca1-ac36-d8811dcc6f53)

Any ideas why the length of **v** and **vt** don't match and how I may get around this?

Thanks so much for your consideration of this so far!

**Reply by HansVansWardikins** [reply]
I have solved this now. I ended up avoiding using `pv.read` until after I have replaced **v** with **uva** to create a new .obj file. I did this by manually parsing the original .obj file, using the extracted vertices and faces as inputs to the `igl.arap` process and then manually writing the resultant **uva** array into a new .obj file with the previously extracted **vt** and **f**. The new file can then be read into Pyvista no problem. Basically what I had previously done with Excel! 

Man I love learning a bit of coding, its amazing what you can do with a bit of googling!

**Reply by darikg** [reply]
Nice! The one thing I was going to suggest was converting directly between pv and igl representations to avoid reading the obj file twice from two different readers.

From igl to pv:
```python
v, f = igl.read(...)
mesh = pv.PolyData.from_regular_faces(v, f)  # in newer pyvista versions
mesh = pv.PolyData(v, faces=np.hstack([np.full((len(f), 1), fill_value=3, dtype='int'), f]))  # if from_regular_faces isn't availabe
```

The other way:
```python
mesh = pv.read(...)
v, f = mesh.points, mesh.regular_faces  # in newer pyvista versions
v, f = mesh.points, mesh.faces.reshape(-1, 4)[:, 1:]  # if regular_faces isn't available
```
